pytensor/tensor/rewriting/blockwise.py

```diff
@@ -46,6 +63,134 @@ def local_useless_unbatched_blockwise(fgraph, node):
 
 # Avoid redundant cases early on for Ops whose default form is not Blockwised
 @register_canonicalize
-@node_rewriter(tracks=[_matrix_matrix_matmul])
+@register_stabilize
+@register_specialize
+@node_rewriter(tracks=[_matrix_matrix_matmul, Blockwise])
 def local_eager_useless_unbatched_blockwise(fgraph, node):
-    return local_useless_unbatched_blockwise.fn(fgraph, node)
+    if (
+        node.op == _matrix_matrix_matmul
+        # These Ops can't always be trivially vectorized at runtime,
+        # since their inputs may dictate non-rectangular shapes.
+        or isinstance(
+            node.op.core_op,
+            (Alloc, ARange, Subtensor, AdvancedSubtensor, AdvancedIncSubtensor),
+        )
+    ):
+        return local_useless_unbatched_blockwise.fn(fgraph, node)
+
+
+def _squeeze_left(x, stop_at_dim: Optional[int] = None):
+    """Squeeze any leading dims of `x` until a real dim or `stop_at_dim` (if not None) is reached."""
+    x_dims = x.type.broadcastable
+    squeeze_ndim = len(x_dims) if all(x_dims) else x_dims.index(False)
+    if stop_at_dim is not None:
+        squeeze_ndim = min(squeeze_ndim, stop_at_dim)
+    if squeeze_ndim == 0:
+        return x
+    return x.squeeze(axis=tuple(range(squeeze_ndim)))
+
+
+@register_specialize("shape_unsafe")
+@node_rewriter([Blockwise])
+def local_blockwise_alloc(fgraph, node):
+    """Push Allocs from the inputs to the output of Blockwise Ops.
+
+    BOp = Blockwise(Op, signature="(x),(x)->(x)")
+    BOp(vector, alloc(vector, 10, 5)) -> alloc(BOp)(vector, vector), 10, 5)
+    BOp(vector, alloc(scalar, 10, 5)) -> alloc(BOp)(vector, alloc(scalar, 5), 10, 5)
+    BOp(matrix, alloc(vector, 10, 5)) -> BOp(matrix, vector)
```

Comment:
I'm guessing that this function is the source of the major speedups in this PR -- I have it in my head that the slowdown was due to alloc operations being blockwise'd. Is that right? If so, why did it cause such terrible graphs?

Reply:
For one thing it was causing huge constants during constant fold.

But more importantly it was causing slow repeated Blockwise operations like Arange with the same inputs. Basically broadcasting before a blockwise means doing the same computation batched number of times which is just silly.

Remember that Blockwise does not have a C implementation so it's just a python loop in np.vectorize. We want to get rid of as many as possible which is what this PR mostly does. Any Blockwise we can avoid in the final function is a win